### PR TITLE
Fix the wire order of state prep operations in `default.qubit`

### LINKED ISSUE #4779: QNodes with `default.qubit` and `default.qubit.legacy` give different results
### Expected behavior

They should give the same results.

### Actual behavior

They give different results.

### Additional information

Seems to be that BasisState followed by some (? — RandomLayers and some instances of QubitUnitary ) operation give different results.

### Source code

```shell
import pennylane as qml
import pennylane.numpy as np

new_device = qml.device("default.qubit", wires=2)
legacy_device = qml.device("default.qubit.legacy", wires=2)

def random_unitary():
    weights = np.array([[0.1, -2.1, 1.4]])
    qml.RandomLayers(weights=weights, wires=0)

def circuit(state):
    qml.BasisState(state, wires=1) 
    random_unitary()
    # getting rid of BasisState and/or random_unitary produces no discrepancy

@qml.qnode(new_device)
def circuit_new_device(state):
    circuit(state)
    return qml.state()

@qml.qnode(legacy_device)
def circuit_legacy_device(state):
    circuit(state)
    return qml.state()

res_new_device = circuit_new_device(np.array([1]))
res_legacy_device = circuit_legacy_device(np.array([1]))

print(res_new_device == res_legacy_device) # [False False False False]
```

### Tracebacks

```shell
N/A
```

### System information

```shell
Name: PennyLane
Version: 0.33.0
Summary: PennyLane is a Python quantum machine learning library by Xanadu Inc.
Home-page: https://github.com/PennyLaneAI/pennylane
Author:
Author-email:
License: Apache License 2.0
Location: /Users/isaac/.virtualenvs/pennylane-stable/lib/python3.9/site-packages
Requires: appdirs, autograd, autoray, cachetools, networkx, numpy, pennylane-lightning, requests, rustworkx, scipy, semantic-version, toml, typing-extensions
Required-by: PennyLane-Lightning

Platform info:           macOS-14.1-x86_64-i386-64bit
Python version:          3.9.14
Numpy version:           1.23.5
Scipy version:           1.10.0
Installed devices:
- default.gaussian (PennyLane-0.34.0.dev0)
- default.mixed (PennyLane-0.34.0.dev0)
- default.qubit (PennyLane-0.34.0.dev0)
- default.qubit.autograd (PennyLane-0.34.0.dev0)
- default.qubit.jax (PennyLane-0.34.0.dev0)
- default.qubit.legacy (PennyLane-0.34.0.dev0)
- default.qubit.tf (PennyLane-0.34.0.dev0)
- default.qubit.torch (PennyLane-0.34.0.dev0)
- default.qutrit (PennyLane-0.34.0.dev0)
- null.qubit (PennyLane-0.34.0.dev0)
- lightning.qubit (PennyLane-Lightning-0.33.1)
```

### Existing GitHub issues

- [X] I have searched existing GitHub issues to make sure the issue does not already exist.

### PR DESCRIPTION
**Pull request by albi3ro**
Fixes #4779 . [sc-49381]

When creating an initial state, `get_final_state` was using `tape.op_wires` for the wire order, even though the simulation always occurred using the wire order `(0,1,2,3,...)`.  Sometimes `tape.op_wires` would have a different order, leading to a mismatch.

Now we use `sorted(circuit.op_wires)`, so they will always be in the monotonically increasing integers used by the simulation.

### CODE REVIEW

File: doc/releases/changelog-dev.md [hunk 1]
```diff
@@ -44,6 +44,9 @@
   is a `Hamiltonian`.
   [(#4768)](https://github.com/PennyLaneAI/pennylane/pull/4768)
 
+* In `defualt.qubit`, initial states are now initialized with the simulation's wire order, not the circuit's
+  wire order.
```
Comment (albi3ro) [inline doc/releases/changelog-dev.md:48]:
```suggestion
  wire order.
  [(#4781)](https://github.com/PennyLaneAI/pennylane/pull/4781)
```

File: doc/releases/changelog-dev.md [hunk 2]
```diff
@@ -44,6 +44,10 @@
   is a `Hamiltonian`.
   [(#4768)](https://github.com/PennyLaneAI/pennylane/pull/4768)
 
+* In `defualt.qubit`, initial states are now initialized with the simulation's wire order, not the circuit's
```
Comment (timmysilv) [inline doc/releases/changelog-dev.md:47]:
```suggestion
* In `default.qubit`, initial states are now initialized with the simulation's wire order, not the circuit's
```

**Review by timmysilv — APPROVED** [review]
oops. nice find, thanks for fixing this!

**Review by isaacdevlugt — APPROVED** [review]
Thanks @albi3ro !

File: doc/releases/changelog-dev.md [hunk 3]
```diff
@@ -44,6 +44,10 @@
   is a `Hamiltonian`.
   [(#4768)](https://github.com/PennyLaneAI/pennylane/pull/4768)
 
+* In `default.qubit`, initial states are now initialized with the simulation's wire order, not the circuit's
```
Comment (trbromley) [inline doc/releases/changelog-dev.md:47]:
```suggestion
* In `default.qubit`, initial states are now initialized with the simulator's wire order, not the circuit's
```

File: tests/devices/default_qubit/test_default_qubit.py [hunk 4]
```diff
@@ -317,6 +317,17 @@ def circuit(p):
...
+        """Test that the wire order is correct with a basis state if the tape wires have a non standard order."""
+
+        dev = DefaultQubit()
+
+        tape = qml.tape.QuantumScript([qml.BasisState([1], wires=1), qml.PauliZ(0)], [qml.state()])
+
+        expected = np.array([0, 1, 0, 0], dtype=np.complex128)
+        res = dev.execute(tape)
+        assert qml.math.allclose(res, expected)
+
```
Comment (trbromley) [inline tests/devices/default_qubit/test_default_qubit.py:330]: Do we also have tests for custom wire labels? E.g., strings or just non-consecutive integers?

Comment (isaacdevlugt) [reply tests/devices/default_qubit/test_default_qubit.py:330]: Good catch!

```
from pennylane.devices import DefaultQubit

dev = DefaultQubit()

tape = qml.tape.QuantumScript([qml.BasisState([1], wires="A"), qml.PauliZ("C")], [qml.state()])

expected = np.array([0, 1, 0, 0], dtype=np.complex128)
res = dev.execute(tape)
```

```
>>> res
array([ 0.+0.j, -0.+0.j,  1.+0.j, -0.+0.j])
```

Comment (albi3ro) [reply tests/devices/default_qubit/test_default_qubit.py:330]: So this same kind of problem won't occur in those cases. If the tape wires are `(2,0)` or `("b", "a")`, then they will get mapped to `(0,1)` at the beginning.

The problem is that the wire order `(1,0)` is not mapped to `(0,1)`, but left as is.

Comment (albi3ro) [reply tests/devices/default_qubit/test_default_qubit.py:330]: @isaacdevlugt  I'm a little confused as to why the "expected" is `np.array([0, 1, 0, 0], dtype=np.complex128)` in this case.


The above example is equivalent to:
```
>>> tape = qml.tape.QuantumScript([qml.BasisState([1], wires=0), qml.PauliZ(1)], [qml.state()])
>>> dev.execute(tape)

array([ 0.+0.j, -0.+0.j,  1.+0.j, -0.+0.j])
```

which is exactly the same.

**Pull request merged**